### PR TITLE
Drop support for EOL versions & test against newer ones

### PR DESCRIPTION
**Pull request by sambostock**
**The most important changes made in this PR are that it drops support for the following:**
- **Ruby < 2.6**
  _Note than 2.6 EOL is scheduled for 31 Mar 2022_
  - 2.5 EOL was 31 Mar 2021
  - 2.4 EOL was 31 Mar 2020
- **Active Support < 5.2**
  - 5.1 EOL was 25 Aug 2019
  - 5.2 EOL was 09 Apr 2018
  - 4.2 EOL was 27 Apr 2017

Several related changes are made:
- We now also test against Ruby 3.0 & 3.1
- We now also test against Active Support 6.1, 7.0, & edge (`rails/rail:main`)
- Multiple gemfiles and gemspec are cleaned up
- All development dependencies are updated

### CODE REVIEW

**Comment by sambostock** [comment]
[This failure](https://github.com/Shopify/deprecation_toolkit/runs/5457731712?check_suite_focus=true) seems to occur on only the _Active Support 7.0_ CI runs, but be fixed in the _edge_ runs.

```
...activesupport-7.0.2.2/lib/active_support/xml_mini.rb:184:in `current_thread_backend':
   uninitialized constant ActiveSupport::XmlMini::IsolatedExecutionState (NameError)

        IsolatedExecutionState[:xml_mini_backend]
```

Using `git bisect` on `rails/rails`, I've identified https://github.com/rails/rails/pull/44340 as fixing it. However, that PR mentions

> For better or worse, this change will incidentally fix the instances of people requiring some individual files from Active Support without first loading active_support.rb that, while not officially supported, were previously working and recently started failing. (https://github.com/rails/rails/issues/43851 https://github.com/rails/rails/pull/43852 https://github.com/rails/rails/issues/43889 https://github.com/rails/rails/issues/43908 https://github.com/rails/rails/issues/44336) This is not a reversal of that supported-usage policy, and a future change could re-break it at any time.

Based on that, we must have a usage of an Active Support internal file without `require "active_support"` first. It would probably make sense to track it down and address it before merging this, rather than skipping the 7.0 entries in the test matrix, but that is always an option. Note that (as far as I can tell), this is revealing an existing bug, not introducing one (though I may be wrong).

**Comment by sambostock** [comment]
The issue is that #50 added https://github.com/Shopify/deprecation_toolkit/blob/0bfe3782acff87e74ce7ff17792de53ce8b98e18/lib/deprecation_toolkit.rb#L38 near the bottom of the file, whereas if needs to be the first thing we do, before even https://github.com/Shopify/deprecation_toolkit/blob/0bfe3782acff87e74ce7ff17792de53ce8b98e18/lib/deprecation_toolkit.rb#L3 because the task code includes https://github.com/Shopify/deprecation_toolkit/blob/0bfe3782acff87e74ce7ff17792de53ce8b98e18/lib/tasks/ci_recorder.rake#L5 and as described in my previous comment, requiring pieces of Active Support before requiring the main file first is unsupported behaviour.

File: CHANGELOG.md [hunk 1]
```diff
@@ -2,6 +2,8 @@
 
 ## master (unreleased)
```
Comment (rochlefebvre) [inline CHANGELOG.md:3]: Nit: now that `master` has been renamed as `main`, let's update the changelog heading as well.

**Review by rochlefebvre — APPROVED** [review]
Nice update and nice cleanup! I learned a few tricks for managing several versioned gemfiles.

File: CHANGELOG.md [hunk 2]
```diff
@@ -1,6 +1,9 @@
 # Change log
 
-## master (unreleased)
+## main (unreleased)
+
+* [#58](https://github.com/Shopify/deprecation_toolkit/pull/58): Drop support for Ruby < 2.6 & Active Support < 5.2. (@sambostock)
+* [#58](https://github.com/Shopify/deprecation_toolkit/pull/58): `require 'active_support'` earlier to ensure compatibility with v7. (@sambostock)
```
Comment (sambostock) [inline CHANGELOG.md:6]: I've only documented the user facing changes. Excluded are:

- the changes to CI infrastructure
- the changes to development dependency versions and config

Comment (rafaelfranca) [inline CHANGELOG.md:6]: We can remove the second one. It is not visible. Or we can change to
```suggestion
* [#58](https://github.com/Shopify/deprecation_toolkit/pull/58): Ensure compatibility with Rails 7. (@sambostock)
```

**Pull request merged**